### DISCUSSION TITLE
Зависание скробблинга last.fm

### DISCUSSION THREAD

**Discussion by efremych** [Chimildic/goofy#137] [Помощь с ошибкой] [answered]
Есть ли какой-то способ избежать следующей ошибки или нивелировать ее последствия?

С завидной регулярностью "закисает" скроблинг Spotify в Last FM.
Т.е. в списке подключенных к Last FM приложений Spotify присутствует, никаких действий с моей стороны не производится с настройками.
Но периодически "закисает" скроблинг и история в LastFM не обновляется. Решается отключением и подключением Spotify в настройках Last FM.

Я даже уже начал отслеживать регулярность этого явления. Правда совсем недавно.
Вот прошел ровно месяц и история в Last FM не обновляется.

Может можно как-то автоматизировать отключение-подключение Spotify к Last FM, чтоб это регулярно происходило без моего участия?
Или выход только ставить себе напоминалку, чтобы периодически выполнять переподключение вручную?

Или же, если то, что выше невозможно, может, можно как-то восстанавливать утерянную за время "закисания" историю прослушивания Last FM? Интересует не просто добавление треков в историю, а добавление с датой/временем прослушивания.
Так как крайне мучительно и практически невозможно добавить пропущенные ранее треки, если в историю смогли/успели подтянутся более свежие, последние треки.

**Comment by Chimildic** [comment, marked as answer]
Тоже встречался с зависанием скробблинга. Но не настолько часто, чтобы имело смысл уходить в дебри. 

Spotify заставил Lastfm пересесть на API. Как я понимаю, за меньшем ресурсом последовали и более частые сбои. 
Lastfm видит только то, что дает API. Оно может само дать сбой, возвращать пустой ответ из-за приватного режима или прослушивания без интернета (при появлении интернета максимум 50 последних треков попадут в историю). Более подробно на [форуме](https://support.last.fm/t/spotify-has-stopped-scrobbling-what-can-i-do/3184).

Скорее всего, проблемы не будет при использовании сторонних приложений скробблинга. Можно найти их и на телефон, и на ПК. Возможно и для авто есть. Тогда не важен приватный режим или офлайн. История будет сохранятся локально и скопом отправляться при возможности. По крайней мере, надеюсь такую реализацию сделали.

> автоматизировать отключение-подключение Spotify к Last FM

Скорее нет

> выход только ставить себе напоминалку

Можно отправить письмо на почту из Apps Script при долгом отсутствии обновлений истории. Если на телефоне есть почтовый клиент, будет и уведомление там.

В следующем коде нужно указать пользователя `user` и электронную почту `to`. Период проверки регулируется аргументами функции [rangeDateRel](https://chimildic.github.io/goofy/#/func?id=rangedaterel). `getRecentTracks` со значением 2, поскольку играющий трек не учитывается и удаляется при наличии. При первом запуске из редактора попросит дать разрешение на отправку писем.
```js
function checkLastfmHistory() {
  let tracks = Lastfm.getRecentTracks('user', 2);
  let date = new Date(tracks[0].played_at);
  Filter.rangeDateRel(tracks, 2, 0);
  if (tracks.length > 0) {
    return;
  }

  MailApp.sendEmail({
    to: "name@gmail.com",
    subject: "Завис lastfm",
    htmlBody: `Последняя дата обновления истории в lastfm - ${date.toGMTString()}`,
  });
}
```

> как-то восстанавливать утерянную за время "закисания" историю прослушивания

Для lastfm есть [программы по редактированию скробблов](https://github.com/SHOEGAZEssb/Last.fm-Scrubbler-WPF). Но не могу сказать насколько это удобно при большем количестве и умеют ли они добавлять новые. Плюс lastfm через API более эффективнее сохраняет историю, чем дает Spotify по запросу 50 последних. Поскольку знает о текущем играющем. Источником потерянной истории может выступить только кэш goofy. Как понимаем, там [свои ограничения](https://chimildic.github.io/goofy/#/?id=История-прослушиваний) и явно не все потерянное найдется.

**Reply by efremych** [reply]
> приватный режим

ну, тут, в приниципе логично, что трек не попадет в историю

> или офлайн

Тут тоже замечал потери, но можно сделать скидку, на то, что всё-таки оффлайн.

> Для lastfm есть программы по редактированию скробблов. Но не могу сказать насколько это удобно при большем количестве и умеют ли они добавлять новые. Плюс lastfm через API более эффективнее сохраняет историю, чем дает Spotify по запросу 50 последних. Поскольку знает о текущем играющем. Источником потерянной истории может выступить только кэш goofy. Как понимаем, там свои ограничения и явно не все потерянное найдется.

Я имею в виду именно добавление треков в кэш. Last FM то без проблем сам подтягивает треки, которые были пропущены во время зависания.
А вот если, допустим, у меня установлено добавление 30 последних треков Last FM в историю и я успел прослушать, допустим, 40 треков до того, как Last FM развис. То в кэш-историю не попадут 10 треков, хоть тресни, даже если изменить число добавляемых треков на 100 штук)

> Можно отправить письмо на почту из Apps Script при долгом отсутствии обновлений истории

Как вариант, да, подумаю над этим, спасибо - позволит уменьшить риск успеть прослушать более 30 треков :)

**Reply by Chimildic** [reply]
> имею в виду именно добавление треков в кэш
```js
let tracks = Lastfm.getRecentTracks('login', 100);
Filter.removeTracks(tracks, RecentTracks.get());
RecentTracks.appendTracks('LastfmRecentTracks', tracks);
```

**Reply by efremych** [reply]
> let tracks = Lastfm.getRecentTracks('login', 100);
Filter.removeTracks(tracks, RecentTracks.get());
RecentTracks.appendTracks('LastfmRecentTracks', tracks);

Хм, про это не подумал. Тем более для RecentTracks завел себе свой removeTracks, значит погрешностей особых быть не должно и дубликаты с прошлого дня прослушивания должны остаться в истории.

Надо было сразу спросить, столько вариантов решения получил. 
Спасибо большое.
